ClassTag in BehaviorInterceptor, #26953

patriknw opened the pull request:
<!--
# Pull Request Checklist

* [ ] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [ ] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [ ] Have you updated the documentation?
* [ ] Have you added tests for any changed functionality?
-->
## Purpose

Make interceptors more safe for unknown message types. Avoiding ClassCastException. Solves the problems when combining EventSourcedBehavior with other interceptors such as `widen` and `withMdc`.

## References

Refs #25887
Refs #26953

Related to https://github.com/akka/akka/pull/27135

## Changes

* Always be explicit about what message types an interceptor can handle, to avoid
  ClassCastException if another message type is passing. That may happen when
  the inner behavior understands other messages than it says in it's declared
  behavior type, for example EventSourcedBehaviorImpl

## Background Context

Several issues with EventSourcedBehavior when combining with an outer interceptor.

patriknw committed:
ClassTag in BehaviorInterceptor

* Always be explicit about what message types an interceptor can handle, to avoid
  ClassCastException if another message type is passing. That may happen when
  the inner behavior understands other messages than it says in it's declared
  behavior type, for example EventSourcedBehaviorImpl

patriknw committed:
Add third, middle, type parameter to BehaviorInterceptor

* Which is the message type the interceptor cares about
* Otherwise it's weird that an interceptor says that it only cares
  about a subtype of `O` but it must anyway deal with `O` in the
  `aroundReceive` method
* This adds some repetition for simple interceptors with same types,
  and if that is seen as problem we could add another base class
  for those. I only see BehaviorInterceptor as power user API so
  maybe doesn't matter if it's verbose?

patriknw committed:
Minimized failing tests

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/InterceptSpec.scala:
```diff
@@ -35,9 +35,53 @@ object InterceptSpec {
       other.isInstanceOf[SameTypeInterceptor]
   }
+
+  // This is similar to how EventSourcedBehavior is implemented
```
This a good starting point that illustrates the problems. See new tests "Protocol transformer interceptor" below.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala:
```diff
@@ -14,30 +17,35 @@ import akka.annotation.{ DoNotInherit, InternalApi }
  * @tparam I The inner message type - the type of message the wrapped behavior accepts
  */
-abstract class BehaviorInterceptor[O, I] {
+abstract class BehaviorInterceptor[O, M <: O, I](val interceptMessageClass: Class[M]) {
```
Isn't this still problematic for event sourced where `M` isn't `<: O`?

What about dropping the type params entirely and thinking of interceptor as super low level (which it is), shouldn't that allow us to not have the extra param/classtag everywhere?

patriknw replied:
This works fine for interceptor in EventSourcedBehavior because that would be [Any, Any, InternalProtocol] and an outer interceptor would be [Command, C <: Command, Command]

By dropping the type param, you mean that an interceptor would always have to handle Any? I don’t think that is good for the more specific interceptors such as withMdc.

johanandren replied:
How would a user interceptor know that it is `Any`, they think it is `TheirCommand`

I was thinking that it would push the decision if a `Class`-instance was needed to the specific interceptor where it actually was needed.

johanandren replied:
I think there may be something more hiding here, `intercept(Interceptor[O, M, I])(Behavior[I])` turns `Behavior[I]` to a `Behavior[O]` but it only actually changes `M` which means that unless `M == O` or the single subtype there is a set of messages that will be passed to the inner behavior and potentially cause a classcast there?

patriknw replied:
You are right. `_ < O` was wrong thinking, and therefore also the third type parameter. I have reverted that so now an interceptor must handle all `O`.

hepin1989 commented on akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala:
```diff
@@ -14,30 +17,35 @@ import akka.annotation.{ DoNotInherit, InternalApi }
+ *           applying the interceptor only to certain message types.
+ *           If the message is not of this class or a subclass thereof it will
+ *           bypass the interceptor and be continue to the inner behavior untouched.
  * @tparam I The inner message type - the type of message the wrapped behavior accepts
```
`Outer` and `Inner` are better than `O` and `I`.

patriknw replied:
ok, I'll change that

patriknw replied:
fixed

patriknw committed:
Revert "Add third, middle, type parameter to BehaviorInterceptor"

This reverts commit e8d64e7edeffd066c1e06dd0661f565f743c52e5.

patriknw committed:
remove `_ <: O`, it must handle all `O`

* Otherwise a subclass of O might not be transformed to
  I and causing unexpected ClassCastExc in the inner behavior because
  it might know nothing of a X <: O
* Changed PoisonPillInterceptor to only intercept signals
  and therefore it doesn't need need the class tag

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala:
```diff
@@ -14,22 +17,28 @@ import akka.annotation.{ DoNotInherit, InternalApi }
+ *                              If `null` then all messages bypass the interceptor and only start and
+ *                              signals are intercepted, i.e. aroundReceive for `msg: O` is not invoked.
+ *                              Using `null` for interceptMessageClass is only safe if `O` and `I` are
+ *                              the same.
```
Instead of documenting that `null` is supported we can have a separate `BehaviorSignalInterceptor[M]` that should be used to intercept signals only. We use that internally in the `PoisonPillInterceptor`.

patriknw replied:
Added separate `BehaviorSignalInterceptor` in 428548d
Pretty nice.

patriknw committed:
rename type params to Outer and Inner

patriknw committed:
separate BehaviorSignalInterceptor

* which only intercepts signals and messages bypass,
  e.g. PoisonPillInterceptor
* also made aroundSignal optional to override in BehaviorInterceptor

patriknw committed:
Add test for interceptors combined with EventSourcedBehavior

patriknw commented:
also refs https://github.com/akka/akka/issues/26953

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:
```diff
@@ -141,7 +141,7 @@ private[akka] final case class MonitorInterceptor[T](actorRef: ActorRef[T]) exte
  */
 @InternalApi
-private[akka] final case class LogMessagesInterceptor[T](opts: LogOptions) extends BehaviorInterceptor[T, T] {
+private[akka] final case class LogMessagesInterceptor[T: ClassTag](opts: LogOptions) extends BehaviorInterceptor[T, T] {
```
Does this really have to have the class tag, can't it just be `Any` since all it does is logging it?

patriknw replied:
good idea, removed it there, in 0abe9eb867406a6e7fd5f42055cbf675cc37d15d

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -48,7 +53,9 @@ import akka.util.unused
-    extends BehaviorInterceptor[O, I] {
+    extends BehaviorInterceptor[Any, I] {
+
+  // FIXME test that supervision works also when exception from a non-O message type, like InternalProtocol in EventSourcedBehavior
```
Add such test or ticketify

patriknw replied:
added test in a1cd6d4fbd96a95db5ee6332a4e607ea6282375d

supervision interceptors are of type Any so intercepting everything

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala:
```diff
@@ -183,24 +183,24 @@ object Behaviors {
    */
-  def monitor[T](monitor: ActorRef[T], behavior: Behavior[T]): Behavior[T] =
-    scaladsl.Behaviors.monitor(monitor, behavior)
+  def monitor[T](interceptMessageClass: Class[_ <: T], monitor: ActorRef[T], behavior: Behavior[T]): Behavior[T] =
```
I never remember, these type bounds work in Java because they become `<? extends T>`?

patriknw replied:
It could have been useful for `monitor` as a filter for subclass of `T`, but I have removed it in 767311c22d8af483304f621532bd8072530b1179 since it seems it was mostly added by accident. If we would have such a feature it should probably be better as  predicate function.

johanandren commented on akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorInterceptorSpec.scala:
```diff
@@ -0,0 +1,119 @@
+
+    "be possible to combine with widen" in {
+      // EventSourcedBehaviorImpl should use a plain BehaviorInterceptor instead of widen
+      pending // FIXME #25887
```
Isn't that the issue this PR is meant to address?

patriknw replied:
This is only the first step for solving several things. To be able to use widen with EventSourcedBehavior we need to use a plain BehaviorInterceptor instead of widen in EventSourcedBehaviorImpl, and that is a good/simple change.

Don't want to add more to this PR, since it already became big. Better with focused reviews of the follow up parts.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:
```diff
@@ -74,9 +76,10 @@ private[akka] final class InterceptorImpl[O, I](
 
   override def receive(ctx: typed.TypedActorContext[O], msg: O): Behavior[O] = {
-    val interceptMessageType = interceptor.interceptMessageType
+    // TODO performance optimization could maybe to avoid isAssignableFrom if interceptMessageClass is Class[Object]?
```
I think it would also be safe to have state in this class, so perhaps we could also cache when a type has been deemed assignable the first time? (If it shows up in benches I guess).

patriknw replied:
the overhead of such cache lookup would probably be higher than isAssignableFrom (which is essentially an instanceof), but let's investigate if that really becomes a problem

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/InterceptorImpl.scala:
```diff
@@ -74,9 +76,10 @@ private[akka] final class InterceptorImpl[O, I](
+    val interceptMessageClass = interceptor.interceptMessageClass
     val result =
-      if (interceptMessageType == null || interceptMessageType.isAssignableFrom(msg.getClass))
+      if ((interceptMessageClass ne null) && interceptor.interceptMessageClass.isAssignableFrom(msg.getClass))
```
This has changed from before, now `null` does _not_ intercept, while before it meant intercept all messages, was that intentional?

patriknw replied:
yes, intentional. The previous `null`, which was also the default was unsafe as we have seen.

The new `null` is needed for when the interceptor shouldn't intercept messages but only signals. That is provided as end user api via `BehaviorSignalInterceptor`

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala:
Why is `BoxedType` needed?

patriknw replied:
there was some test (for widen I think) that failed for Int vs Integer

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala:
Should we maybe add a note that it is a power user feature and not the first thing you should reach for (given recent discussion)?

patriknw replied:
Ok, we can add something more to above "Many built-in intercepting behaviors are provided through factories in the respective `Behaviors`."

patriknw replied:
added in 362ac824d038d83f7a96be466b02581d2d59b4f3

johanandren reviewed (commented):
Having the classtag/class everywhere makes me a bit sad, but we have discussed this so much and not found any other solution, so I guess it's just how it has to be.

Mostly looks good, some sprinkled questions and small things.

patriknw committed:
ClassTag not needed for LogMessagesInterceptor

* since it can handle Any

patriknw committed:
test supervision of different message type

patriknw committed:
clarify low level

patriknw committed:
remove _ <: T for the interceptMessageClass

* could have been useful for filtering out a certain subtype of messages in `monitor`

chbatey commented on akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala:
do we need to override this, looks the same impl

patriknw replied:
thanks, good catch

chbatey commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/InterceptSpec.scala:
```diff
@@ -28,16 +30,48 @@ object InterceptSpec {
+          ctx: TypedActorContext[Any],
+          msg: Any,
+          target: BehaviorInterceptor.ReceiveTarget[InternalProtocol]): Behavior[InternalProtocol] = {
+        val wrapped = msg match {
```
this match would fail if another message comes e.g. a timer?

patriknw replied:
The purpose of this is to test the same as we use in EventSourcedBehaviorImpl. There it's know that the incoming messages can either be ExternalProtocol or Command. Other unknown messages will blow up (but that would be a bug if other messages were sent to it). Doesn't matter if they blow up in the interceptor or in the receiveMessage of the inner Behavior.

Timer messages are not implemented with interceptors any more, they are unwrapped at the outer layer in ActorAdapter.

johanandren reviewed (approved):
LGTM (with Chris's comments addressed)

patriknw committed:
more docs for interceptMessageClass param and ClassTag

patriknw committed:
remove O type parameter in supervision

patriknw committed:
remove extra setup for RestartSupervisor, already factory

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -33,11 +33,16 @@ import akka.util.unused
       case r: RestartOrBackoff =>
-        Behaviors.intercept[T, T](() => new RestartSupervisor(initialBehavior, r))(initialBehavior)
+        Behaviors.setup { _ =>
+          // deferred to make sure supervisor instance not shared among instances
```
good catch, must be leftover from before it was a factory, fixed in ff96a0619851b17151e5bd5303c612d912f7723d

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/Supervision.scala:
```diff
@@ -48,7 +53,7 @@ import akka.util.unused
 @InternalApi
 private abstract class AbstractSupervisor[O, I, Thr <: Throwable](strategy: SupervisorStrategy)(
```
yes, not used, removed in 4368db43ca7f8b20ccb13fb9eb0957088e177192

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala:
```diff
@@ -276,8 +276,11 @@ object Behaviors {
-  def widened[T, U](behavior: Behavior[T], selector: JFunction[PFBuilder[U, T], PFBuilder[U, T]]): Behavior[U] =
-    BehaviorImpl.widened(behavior, selector.apply(new PFBuilder).build())
+  def widened[T, U](
+      interceptMessageClass: Class[T],
```
added in 5b57adf61d5bc5db680b927223c3b98f7faadda6

patriknw committed:
mention in migration guide

patriknw commented:
also added note in migration guide: f4d0e45b9879ad8a75c46955842563634396ca42
